internal/provisioner/helm_utils.go

```diff
@@ -0,0 +1,225 @@
+package provisioner
+
+import (
+	"bytes"
+	"context"
+	"os/exec"
+	"time"
+
+	"github.com/mattermost/mattermost-cloud/internal/tools/k8s"
+	"github.com/mattermost/mattermost-cloud/internal/tools/kops"
+	"github.com/mattermost/mattermost-cloud/model"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+// helmDeployment deploys Helm charts.
+type helmDeployment struct {
+	chartDeploymentName string
+	chartName           string
+	namespace           string
+	setArgument         string
+	valuesPath          string
+
+	cluster         *model.Cluster
+	kopsProvisioner *KopsProvisioner
+	kops            *kops.Cmd
+}
+
+func installHelm(kops *kops.Cmd, logger log.FieldLogger) error {
+	logger.Info("Installing Helm")
+
+	err := helmSetup(logger, kops)
+	if err != nil {
+		return err
+	}
+
+	wait := 120
+	logger.Infof("Waiting up to %d seconds for helm to become ready...", wait)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(wait)*time.Second)
+	defer cancel()
+	err = waitForHelmRunning(ctx, kops.GetKubeConfigPath())
+	if err != nil {
+		return err
+	}
+
+	logger.Info("Updating all Helm repos.")
+	return helmRepoUpdate(logger)
+}
+
+func (d *helmDeployment) Update() error {
+	logger := d.kopsProvisioner.logger.WithField("helm-update", d.chartName)
+
+	logger.Infof("Refreshing helm chart %s -- may trigger service upgrade", d.chartName)
+	err := upgradeHelmChart(*d, d.kops.GetKubeConfigPath(), logger)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Create will install a given HelmDeployment in the cluster specified in that object.
+func (d *helmDeployment) Create() error {
+	logger := d.kopsProvisioner.logger.WithField("helm-create", d.chartName)
```

Comment:
This is better, but isn't quite what I think we want.

We normally create loggers when the provisioner starts managing a cluster (create, upgrade, etc.)

When it does so, it slowly appends fields to the logger (IDs, states, etc) so we should pass those loggers in when we need to log something on one of the utilities instead of creating a new logging object.

Here is an example of a logger we would want pass in as it will contain the cluster ID and other info: https://github.com/mattermost/mattermost-cloud/blob/master/internal/provisioner/kops_provisioner_cluster.go#L85

I could see doing this one of two ways:
1. Change the interface to accept loggers for everything as an argument
2. Require a logger to be passed in when creating the utilities

Either should be fine for now, so go ahead and decide which one makes more sense to you.

Reply:
OK I have required a logger be passed in while creating the utilities. I think I understand the pattern here -- we accumulate fields on the log entries similar to the way that errors bubble up and accumulate increasingly broad error messages.

In places where there is no obvious metadata that should accompany the field name, when creating the new logger, I have provided a UUID as a value, so that we can trace unique operations in log messages. This was a bit of a judgement call, so feel free to suggest something else if you are not a fan of this pattern, either generally or for specific places where I am creating logger objects.